sematic/external_resource.py

```diff
@@ -0,0 +1,229 @@
+    state:
+        The discrete state that the resource is in
+    message:
+        A human-readable message about why the resource entered this state
+    last_update_epoch_time:
+        The last time this status was checked against the actual external resource
+    """
+
+    state: ResourceState
+    message: str
+    last_update_epoch_time: int
+
+
+@dataclass(frozen=True)
+class ExternalResource(AbstractExternalResource):
+    """Represents a resource tracked by Sematic for usage in Sematic funcs.
+
+    Examples of possible external resources include small data processing
+    clusters or distributed training clusters.
+
+    Specific external resources should inherit from this class. They may add their
+    own dataclass attributes and must implement the _do_activate, _do_deactivate, and
+    _do_update methods on this class.
+
+    This class represents immutable objects. As with immutable strings, the
+    preferred pattern is to replace instances with updated copies rather
+    than modify in-place. This allows for better control over validation of
+    objects and state transitions.
+
+    Attributes
+    ----------
+    id:
+        A UUID representing this particular resource instance
+    status:
+        Information about the current status of the resource
+    """
+
+    id: str = field(default_factory=lambda: uuid.uuid4().hex)
+    status: ResourceStatus = field(
+        default_factory=lambda: ResourceStatus(
+            state=ResourceState.CREATED,
+            message="Resource has not been activated yet",
+            last_update_epoch_time=int(time.time()),
+        )
+    )
+
+    def __post_init__(self):
+        try:
+            uuid.UUID(hex=self.id)
+        except ValueError:
+            raise ValueError(f"ExternalResource had an invalid uuid: '{self.id}'")
+        if not isinstance(self.status, ResourceStatus):
+            raise ValueError(f"ExternalResource had invalid status: '{self.status}'")
+
+    @final
+    def activate(self, is_local: bool) -> "ExternalResource":
+        """Perform the initialization of the resource.
+
+        Parameters
+        ----------
+        is_local:
+            True if a local version of the resource should be initialized, False
+            otherwise. If the resource can't be initialized locally, can raise
+            NotImplementedError. If is_local is False, this logic will be executed
+            on the server.
+
+        Returns
+        -------
+        An updated copy of the object
+        """
+        logger.debug("Activating %s", self)
+        updated = self._do_activate(is_local)
+        self._validate_transition(updated)
+        if updated.status.state != ResourceState.ACTIVATING:
+            raise IllegalStateTransitionError(
+                "Calling .activate() should always leave the resource in the "
+                "ACTIVATING state."
+            )
+        return updated
+
+    def _do_activate(self, is_local: bool) -> "ExternalResource":
```

Comment:
The resource can do it via the `api_client`. The resource knows its own id, which the server also knows about. So an `ExternalResource` could do something like this:
```
def _do_activate(self, is_local: bool) -> "MyResource":
    if is_local:
        # ...
        return new_value
    else:
        return api_client.activate_resource(self.id)
```

The server will already know about the resource by the time `activate` is called because of the `_register_external_resources` call in the resolvers. The cloud resolver will use that hook to save the resource to the server